sdk/javascript/client/stream.js

```diff
@@ -47,42 +48,64 @@ const _generateId = () =>
 class Stream {
   constructor (url) {
     this._url = url
-    this._socket = zmq.socket('dealer')
-    this._socket.setsockopt('identity',
-                            Buffer.from(this.constructor.name + uuid(), 'utf8'))
+    this._identity = Buffer.from(this.constructor.name + uuid(), 'utf8')
+
     this._futures = {}
   }
 
-  connect () {
+  connect (onConnectCb) {
+    this._onConnectCb = onConnectCb
+    this._socket = zmq.socket('dealer')
+    this._socket.setsockopt('identity', this._identity)
     this._socket.connect(this._url)
+    this._socket.on('disconnect', (fd, endpoint) => this._handleDisconnect())
+    this._socket.monitor(500, zmq.ZMQ_EVENT_DISCONNECTED)
+    this._onConnectCb()
   }
 
   close () {
+    this._socket.setsockopt(zmq.ZMQ_LINGER, 0)
+    this._socket.unmonitor()
     this._socket.close()
+    this._socket = null
+  }
+
+  _handleDisconnect () {
+    console.log(`Disconnected from ${this._url}`)
+    this.close()
+    Object.keys(this._futures).forEach((correlationId) => {
+      this._futures[correlationId].reject(
+        new ValidatorConnectionError('the connection to the validator was lost'))
+    })
+
+    this.connect(this._onConnectCb)
   }
 
   send (type, content) {
-    const correlationId = _generateId()
-    let future = new Future()
-    this._futures[correlationId] = future
+    if (this._socket) {
+      const correlationId = _generateId()
+      let deferred = new Deferred()
+      this._futures[correlationId] = deferred
 
-    this._socket.send(_encodeMessage(type, correlationId, content))
+      this._socket.send(_encodeMessage(type, correlationId, content))
 
-    return future
+      return deferred.promise
+    }
   }
 
   sendBack (type, correlationId, content) {
```

Comment:
It currently follows the python and java sdks, but we could update any/all of them.